accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+
+We could also use have only shared ref version of HttpClientHandler but then every optional SDKs would have to create their own copy of proxy logic as described in **Proposal #2**.
+
+**Located in dotnet/runtime repo**
+```c#
+// System.Net.Http.dll - ref for TFM net5.0-ios
+public class HttpClientHandler
+{
+  public HttpClientHandler ()
+  {
+  }
+
+  public long MaxInputInMemory { get; set; }
+
+  public NSUrlSessionHandlerTrustOverrideCallback TrustOverride { get; set; }
+  
+  // More unique or common properties to follow
+}
+```
+
+```c#
+// System.Net.Http.dll
+
+#if iOS
+// The reference for the type comes from Xamarin.iOS.dll
+[assembly: TypeForwardTo (typeof (System.Net.Http.HttpClientHandler)) 
+#elif (android)
+// The reference for the type comes from Xamarin.Android.dll
+[assembly: TypeForwardTo (typeof (System.Net.Http.HttpClientHandler)) 
+#else if ..
+
+#endif
+```
+
+**Disadvantages**
+
+* dotnet/runtime will need to include reference assemblies of all optional components to be able to build ref and impl assemblies.
+* Public HttpClientHandler API shape might be different between TFM as it’s controlled by optional components
+* Updating HttpClientHandler public surface for optional component will require build a new version of dotnet/runtime
+
+**Conclusion**
+
+The big advantage of this proposal is that the default HttpClientHandler would be have the most optimal implementation on every platform (performance and size wise). The availability of only platform specific properties on HttpClientHandler has its pros & cons as well.
+
+### Proposal #5: All HttpHandlers are fully implemented inbox
+
+Implement all platform-specific handlers fully inside dotnet/runtime using platform-specific interop libraries. We would extend low-level interop API which and use it to implement the managed API for HttpClientHandler and other types which are exposed from it.
+
+There is already a foundation for CF interop at [https://github.com/dotnet/runtime/tree/master/src/libraries/Common/src/Interop/OSX](https://github.com/dotnet/runtime/tree/master/src/libraries/Common/src/Interop/OSX). This will need to be extended to include Objective-C bridge for necessary APIs.
+
+We could also expose this Interop layer to Xamarin internally to avoid code duplication as they use the same platform API.
+
+**Disadvantages**
+
+* Implementation of HttpClientHandler might require implementing more dependent type (e.g. Android.Net.SSLSocketFactory)
+* Problems with handling platform-specific exceptions
+* Not all decorations used for Xamarin bindings will be available (e.g. [iOS (7,0)])
+* Starting from scratch for Android
+
+**Conclusion**
+
+No build magic and more shared code prospect, as well as the ability to test HttpClient changes easily across all platforms, are very appealing.
```

Comment:
It won't ease testing as you won't have the (ObjC for XI/XM) runtime. This is creating a new cyclic dependency as you'll need the SDK to create something going back to a different/newer SDK.

Reply:
That's not what the text is saying. There would be no dependency on XI runtime all needed code would exist inside dotnet/runtime.